Update Access Control docs (#719)
OpenZeppelin/cairo-contracts: docs/modules/ROOT/pages/api/access.adoc

:github-icon: pass:[<svg class="icon"><use href="#github-icon"/></svg>]
:AccessControl: xref:AccessControl[AccessControl]
:Ownable: xref:Ownable[Ownable]
:src5: https://github.com/starknet-io/SNIPs/blob/main/SNIPS/snip-5.md[SRC5]
:inner-src5: xref:api/introspection.adoc#ISRC5[SRC5 ID]
:_set_role_admin: xref:#AccessControl-_set_role_admin[_set_role_admin]

= Access Control

[.readme-notice]
NOTE: This document is better viewed at https://docs.openzeppelin.com/contracts/api/access

This directory provides ways to restrict who can access the functions of a contract or when they can do it.

- {Ownable} is a simple mechanism with a single "owner" role that can be assigned to a single account.

> This mechanism can be useful in simple scenarios, but fine grained access needs are likely to outgrow it.

Reviewer: ```suggestion
This mechanism can be useful in simple scenarios, but fine-grained access needs are likely to outgrow it.
```

> - {AccessControl} provides a general role based access control mechanism. Multiple hierarchical roles can be created and

Reviewer: ```suggestion
- {AccessControl} provides a general role-based access control mechanism. Multiple hierarchical roles can be created and
```

assigned each to multiple accounts.

== Authorization

[.contract]
[[Ownable]]
=== `++Ownable++` link:https://github.com/OpenZeppelin/cairo-contracts/blob/release-0.7.0/src/access/ownable/ownable.cairo[{github-icon},role=heading-link]

```javascript
use openzeppelin::access::ownable::Ownable;
```

`Ownable` provides a basic access control mechanism where an account
 (an owner) can be granted exclusive access to specific functions.

This module includes the `assert_only_owner` internal to restrict a function to be used only by the owner.

[.contract-index]
.External Functions
--
.OwnableImpl

* xref:Ownable-owner[`++owner(self)++`]
* xref:Ownable-transfer_ownership[`++transfer_ownership(self, new_owner)++`]
* xref:Ownable-renounce_ownership[`++renounce_ownership(self)++`]
--

[.contract-index]
.Internal Functions
--
.InternalImpl

* xref:Ownable-initializer[`++initializer(self, owner)++`]

> * xref:Ownable-_transfer_ownership[`++_transfer_ownership(self, new_owner)++`]
> * xref:Ownable-assert_only_owner[`++assert_only_owner(self)++`]

Reviewer: ```suggestion
* xref:Ownable-assert_only_owner[`++assert_only_owner(self)++`]
* xref:Ownable-_transfer_ownership[`++_transfer_ownership(self, new_owner)++`]
```
^So that they match the order in the contract. I'd switch their respective descriptions below as well

--

[.contract-index]
.Events
--
* xref:Ownable-OwnershipTransferred[`++OwnershipTransferred(previous_owner, new_owner)++`]
--

[#Ownable-External-Functions]
==== External Functions

[.contract-item]
[[Ownable-owner]]
==== `[.contract-item-name]#++owner++#++(self: @ContractState) → ContractAddress++` [.item-kind]#external#

Returns the address of the current owner.

[.contract-item]
[[Ownable-transfer_ownership]]
==== `[.contract-item-name]#++transfer_ownership++#++(ref self: ContractState, new_owner: ContractAddress)++` [.item-kind]#external#

Transfers ownership of the contract to a new account (`new_owner`).
Can only be called by the current owner.

Emits an xref:Ownable-OwnershipTransferred[OwnershipTransferred] event.

[.contract-item]

> [[Ownable-renounce_ownership--]]

Reviewer: ```suggestion
[[Ownable-renounce_ownership]]
```

==== `[.contract-item-name]#++renounce_ownership++#++(ref self: ContractState)++` [.item-kind]#external#

Leaves the contract without owner. It will not be possible to call
`assert_only_owner` functions anymore. Can only be called by the current owner.

NOTE: Renouncing ownership will leave the contract without an owner,
thereby removing any functionality that is only available to the owner.

[#Ownable-Internal-Functions]
==== Internal Functions

[.contract-item]
[[Ownable-initializer]]
==== `[.contract-item-name]#++initializer++#++(ref self: ContractState, owner: ContractAddress)++` [.item-kind]#internal#

> Initializes the contract setting `owner` as the initial owner.

Reviewer: ```suggestion
Initializes the contract and sets `owner` as the initial owner.
```

Requirements:

- the caller must have ``role``'s admin role.

May emit a {RoleRevoked} event.

[.contract-item]
[[AccessControl-renounce_role]]
==== `[.contract-item-name]#++renounce_role++#++(ref self: ContractState, role: felt252, account: ContractAddress)++` [.item-kind]#external#

Revokes `role` from the calling account.

Roles are often managed via {grant_role} and {revoke_role}. This function's
purpose is to provide a mechanism for accounts to lose their privileges
if they are compromised (such as when a trusted device is misplaced).

If the calling account had been revoked `role`, emits a {RoleRevoked}
event.

Requirements:

- the caller must be `account`.

May emit a {RoleRevoked} event.

[.contract-item]
[[AccessControl-supports_interface]]
==== `[.contract-item-name]#++supports_interface++#++(self: @ContractState, interface_id: felt252) → bool++` [.item-kind]#external#

Returns whether a contract implements a given interface or not.

[#AccessControl-Internal-Functions]
==== Internal Functions

[.contract-item]
[[AccessControl-initializer]]
==== `[.contract-item-name]#++initializer++#++(ref self: ContractState)++` [.item-kind]#internal#

> Initializes the contract by registering the IAccessControl interface ID.

Reviewer: What do you think about linking `IAccessControl interface ID` to the `IAccessControl` section? This way, users will be able to find the interface id when they click on the link

- {Ownable} is a simple mechanism with a single "owner" role that can be assigned to a single account.
This mechanism can be useful in simple scenarios, but fine grained access needs are likely to outgrow it.
- {AccessControl} provides a general role based access control mechanism. Multiple hierarchical roles can be created and
assigned each to multiple accounts.

== Authorization

[.contract]
[[Ownable]]
=== `++Ownable++` link:https://github.com/OpenZeppelin/cairo-contracts/blob/cairo-2/src/access/ownable/ownable.cairo[{github-icon},role=heading-link]

```javascript
use openzeppelin::access::ownable::Ownable;
```

`Ownable` provides a basic access control mechanism where an account
 (an owner) can be granted exclusive access to specific functions.

This module includes the `assert_only_owner` internal to restrict a function to be used only by the owner.

[.contract-index]
.External Functions
--
[.contract-subindex-inherited]
.OwnableImpl

* xref:Ownable-owner[`++owner(self)++`]
* xref:Ownable-transfer_ownership[`++transfer_ownership(self, new_owner)++`]
* xref:Ownable-renounce_ownership[`++renounce_ownership(self)++`]
--

[.contract-index]
.Internal Functions
--

> [.contract-subindex-inherited]

Reviewer 2: To avoid subsections being grayed out, you can remove `[.contract-subindex-inherited]`. I've updated the CSS so that the only difference is the opacity.

Author: No strong defense, but to add the comment, I feel it may look better if we let the subindex grayed (not the list element), because being both black now, looks weird to me in the UI.

I don't have a problem with all being grayed out for consistency anyway, but let me know what do you think @martriay  (this will affect other PRs as well)

Reviewer 3: @ericnordelo not sure what you mean, can you screenshot the two alternatives to compare?

Author: I just realized that if what you meant is to remove all the grayed-out sections (make everything black), that makes sense to me. I was thinking of just putting the first section black, and the rest grayed out. Is that what you meant?

Reviewer 3: yes all of them

Reviewer 3: that looks right

assigned each to multiple accounts.
- {Ownable} is a simpler mechanism with a single owner "role" that can be assigned to a single account.
This simpler mechanism can be useful for quick tests but projects with production concerns are likely to outgrow it.

== Authorization

[.contract]
[[Ownable]]
=== `++Ownable++` link:https://github.com/OpenZeppelin/cairo-contracts/blob/cairo-2/src/access/ownable/ownable.cairo[{github-icon},role=heading-link]

[.hljs-theme-light.nopadding]
```javascript
use openzeppelin::access::ownable::Ownable;
```

Contract module which provides a basic access control mechanism, where
there is an account (an owner) that can be granted exclusive access to
specific functions.

This module provides an internal utility named `assert_only_owner`, which can be applied to your functions
to restrict their use to the owner.

[.contract-index]
.Utilities
--
* xref:Ownable-assert_only_owner[`++InternalImpl::assert_only_owner(self)++`]
--

[.contract-index]
.Functions
--
[.contract-subindex-inherited]
.InternalImpl

* xref:Ownable-initializer[`++initializer(self, owner)++`]

> * xref:Ownable-_transfer_ownership[`++_transfer_ownership(self, new_owner)++`]

Reviewer: I feel like we should list the `InternalImpl` at the end of the `Functions` list. That's how they're listed in the accompanying account docs and the Solidity docs. Wdyt?

Reviewer: In the previous iteration, I think we can keep the initializer at the top of `InternalImpl`...but if we're also throwing the constructor in there, something like this seems fitting:

```
constructor()                                         INTERNALIMPL
initializer()
_internal_func()
...
```

> If we follow the proposal of splitting External Functions and Internal Functions this will be solved as well.

And yeah, if we follow the proposal^, I think the approach you have works as well

Author: > also i don't see the point of having Utilities as a separate section, why not part of Internals?

My thinking is because those will be used more frequently than other internals, and is better to catch the user attention to those first IMO. I'm ok with moving them to internal if you think is not worth.

Author: > in the meantime maybe we can

In the UI, what you are suggesting is to have a section like Utilities just for the constructor? Because that's the only way i can think of for presenting it first, and it seems weird to me to have a section Constructor, with the constructor method alone.

Reviewer 3: > > also i don't see the point of having Utilities as a separate section, why not part of Internals?
> 
> My thinking is because those will be used more frequently than other internals, and is better to catch the user's attention to those first IMO. I'm ok with moving them to internal if you think is not worth it.

i guess we can call them "asserts" or "guards", like modifiers. but anything can be a utility, like `_mint`

Author: Your call on using Asserts or removing the section, let me know to update accordingly.

Author: > In the UI, what you are suggesting is to have a section like Utilities just for the constructor? Because that's the only way i can think of for presenting it first, and it seems weird to me to have a section Constructor, with the constructor method alone.

Let me know what are you suggesting about constructors.

Reviewer 3: > In the UI, what you are suggesting is to have a section like Utilities just for the constructor?

sorry, missed it. yes that's what i was suggesting

Reviewer 3: > Your call on using Asserts or removing the section, let me know to update accordingly.

i'd remove it for consistency with the rest of the sections. we can [add guards in a future iteration](https://github.com/OpenZeppelin/cairo-contracts/issues/745).

[.contract-subindex-inherited]
.OwnableImpl

* xref:Ownable-owner[`++owner(self)++`]
* xref:Ownable-renounce_ownership[`++renounce_ownership(self)++`]
* xref:Ownable-transfer_ownership[`++transfer_ownership(self, new_owner)++`]

--

[.contract-index]
.Events
--
* xref:Ownable-OwnershipTransferred[`++OwnershipTransferred(previous_owner, new_owner)++`]
--

[#Ownable-Utilities]
==== Utilities

[.contract-item]
[[Ownable-assert_only_owner]]
==== `[.contract-item-name]#++assert_only_owner++#++(self: @ContractState)++` [.item-kind]#internal#

> Throws if called by any account other than the owner.

Reviewer 3: I think this is ok. We should normalize this across the docs. https://github.com/OpenZeppelin/cairo-contracts/issues/726#issuecomment-1720164104

Transfers ownership of the contract to a new account (`new_owner`).
Internal function without access restriction.

Emits an xref:Ownable-OwnershipTransferred[OwnershipTransferred] event.

[.contract-item]
[[Ownable-owner]]
==== `[.contract-item-name]#++owner++#++(self: @ContractState) → ContractAddress++` [.item-kind]#external#

Returns the address of the current owner.

[.contract-item]
[[Ownable-renounce_ownership--]]
==== `[.contract-item-name]#++renounce_ownership++#++(ref self: ContractState)++` [.item-kind]#external#

Leaves the contract without owner. It will not be possible to call
`assert_only_owner` functions anymore. Can only be called by the current owner.

NOTE: Renouncing ownership will leave the contract without an owner,
thereby removing any functionality that is only available to the owner.

[.contract-item]
[[Ownable-transfer_ownership]]
==== `[.contract-item-name]#++transfer_ownership++#++(ref self: ContractState, new_owner: ContractAddress)++` [.item-kind]#external#

Transfers ownership of the contract to a new account (`new_owner`).

> Can only be called by the current owner.

Author: I don't think is necessary. Correct me if I'm wrong, but I think this would be relevant only if an Account implements the Ownable extension, and tries to transfer ownership, after setting the owner to 0, by calling the transfer_ownership internally in the execute method (only place where caller can be 0). I don't think we need to explain that trying to transfer the ownership in this case will revert.

> Can only be called by the current owner.

Reviewer: Haha my favorite edge case :) yes, that's a fair point

> Can only be called by the current owner.

Reviewer 3: wait why do we prevent calling from zero? is that the only scenario?

> Can only be called by the current owner.

Reviewer: > something is telling me this was my call...

Team effort! https://github.com/OpenZeppelin/cairo-contracts/pull/604#discussion_r1179863438

> Can only be called by the current owner.

Reviewer 3: at least i'm consistent

Emits an xref:Ownable-OwnershipTransferred[OwnershipTransferred] event.

[#Ownable-Events]
==== Events

[.contract-item]
[[Ownable-OwnershipTransferred]]
==== `[.contract-item-name]#++OwnershipTransferred++#++(previous_owner: ContractAddress, new_owner: ContractAddress)++` [.item-kind]#event#

Emitted when the ownership is transferred.

[.contract]
[[IAccessControl]]
=== `++IAccessControl++` link:https://github.com/OpenZeppelin/cairo-contracts/blob/05429e4fd34a250ce7a01450190c53275e5c1c0b/src/access/accesscontrol/interface.cairo#L10[{github-icon},role=heading-link]

> :grant_role: xref:#IAccessControl-grant_role[grant_role]

Reviewer: Should we include the IACCESS_CONTROL ID for SRC-5 somewhere in here? I think it'd be a nice benefit for users

Reviewer: Another potential spot could be in the `Interface` section in the access page itself

--

[.contract-index]
.Events
--
* xref:IAccessControl-RoleAdminChanged[`++RoleAdminChanged(role, previous_admin_role, new_admin_role)++`]
* xref:IAccessControl-RoleGranted[`++RoleGranted(role, account, sender)++`]
* xref:IAccessControl-RoleRevoked[`++RoleRevoked(role, account, sender)++`]

--

[#IAccessControl-Functions]
==== Functions

[.contract-item]
[[IAccessControl-has_role]]
==== `[.contract-item-name]#++has_role++#++(role: felt252, account: ContractAddress) → bool++` [.item-kind]#external#

Returns `true` if `account` has been granted `role`.

[.contract-item]
[[IAccessControl-get_role_admin]]
==== `[.contract-item-name]#++get_role_admin++#++(role: felt252) → felt252++` [.item-kind]#external#

Returns the admin role that controls `role`. See {grant_role} and
{revoke_role}.

To change a role's admin, use {AccessControl-_setRoleAdmin}.

[.contract-item]
[[IAccessControl-grant_role]]
==== `[.contract-item-name]#++grant_role++#++(role: felt252, account: ContractAddress)++` [.item-kind]#external#

Grants `role` to `account`.

If `account` had not been already granted `role`, emits a {RoleGranted}
event.

Requirements:

> - the caller must have ``role``'s admin role.

Author: I'm leaning towards keeping Requirements on this. I think is easier to write "requirements" semantically in a list, than keeping a "Panics" entry. I like more this:

```
Requirements:

- the caller must have ``role``'s admin role.
```

than this:

```
Panics:

- the caller must have ``role``'s admin role.
```

or this:

```
Panics:

- if the caller doesn't have ``role``'s admin role.
```

Lmk what you think

> - the caller must have ``role``'s admin role.

Reviewer 3: i like _requirements_ better. it's a concept i'm familiar with and it fits documentation. Panic scenarios are also useful but I associate that with runtime problems/exceptions. Conceptually, reqs come first.

Transfers ownership of the contract to a new account (`new_owner`).
Can only be called by the current owner.

Emits an xref:Ownable-OwnershipTransferred[OwnershipTransferred] event.

[#Ownable-Events]
==== Events

[.contract-item]
[[Ownable-OwnershipTransferred]]
==== `[.contract-item-name]#++OwnershipTransferred++#++(previous_owner: ContractAddress, new_owner: ContractAddress)++` [.item-kind]#event#

Emitted when the ownership is transferred.

[.contract]
[[IAccessControl]]
=== `++IAccessControl++` link:https://github.com/OpenZeppelin/cairo-contracts/blob/05429e4fd34a250ce7a01450190c53275e5c1c0b/src/access/accesscontrol/interface.cairo#L10[{github-icon},role=heading-link]

:grant_role: xref:#IAccessControl-grant_role[grant_role]
:revoke_role: xref:#IAccessControl-revoke_role[revoke_role]
:RoleGranted: xref:#IAccessControl-RoleGranted[RoleGranted]
:RoleRevoked: xref:#IAccessControl-RoleRevoked[RoleRevoked]
:RoleAdminChanged: xref:#IAccessControl-RoleAdminChanged[RoleAdminChanged]

[.hljs-theme-light.nopadding]
```javascript
use openzeppelin::access::accesscontrol::interface::IAccessControl;
```

External interface of AccessControl declared to support {src5} detection.

> The {src5} ID for this interface is `0x23700be02858dbe2ac4dc9c9f66d0b6b0ed81ec7f970ca6844500a56ff61751`.

Author: Lmk what you think.

> The {src5} ID for this interface is `0x23700be02858dbe2ac4dc9c9f66d0b6b0ed81ec7f970ca6844500a56ff61751`.

Reviewer 3: like it! would only add a link as mentioned above

> The {src5} ID for this interface is `0x23700be02858dbe2ac4dc9c9f66d0b6b0ed81ec7f970ca6844500a56ff61751`.

Author: I like SRC5 ID. 

> would only add a link as mentioned above

Not sure what link you are referring to.

assigned each to multiple accounts.

== Authorization

[.contract]
[[Ownable]]
=== `++Ownable++` link:https://github.com/OpenZeppelin/cairo-contracts/blob/cairo-2/src/access/ownable/ownable.cairo[{github-icon},role=heading-link]

```javascript
use openzeppelin::access::ownable::Ownable;
```

Contract module which provides a basic access control mechanism, where
there is an account (an owner) that can be granted exclusive access to
specific functions.

This module provides an internal utility named `assert_only_owner`, which can be applied to your functions
to restrict their use to the owner.

[.contract-index]
.Utilities
--
* xref:Ownable-assert_only_owner[`++InternalImpl::assert_only_owner(self)++`]
--

[.contract-index]
.Functions
--
[.contract-subindex-inherited]
.InternalImpl

* xref:Ownable-initializer[`++initializer(self, owner)++`]
* xref:Ownable-_transfer_ownership[`++_transfer_ownership(self, new_owner)++`]

[.contract-subindex-inherited]
.OwnableImpl

* xref:Ownable-owner[`++owner(self)++`]
* xref:Ownable-renounce_ownership[`++renounce_ownership(self)++`]

> * xref:Ownable-transfer_ownership[`++transfer_ownership(self, new_owner)++`]

Author: Everything in a sub-index (OwnableImpl) is grayed out :(, that's a UI thing.

> * xref:Ownable-transfer_ownership[`++transfer_ownership(self, new_owner)++`]

Reviewer 3: i know, just pointing it out. cc/ @frangio

Emitted when `account` is granted `role`.

`sender` is the account that originated the contract call, an admin role
bearer.

[.contract-item]
[[IAccessControl-RoleRevoked]]
==== `[.contract-item-name]#++RoleRevoked++#++(role: felt252, account: ContractAddress, sender: ContractAddress)++` [.item-kind]#event#

Emitted when `account` is revoked `role`.

`sender` is the account that originated the contract call:

- if using `revoke_role`, it is the admin role bearer.
- if using `renounce_role`, it is the role bearer (i.e. `account`).

[.contract]
[[AccessControl]]
=== `++AccessControl++` link:https://github.com/OpenZeppelin/cairo-contracts/blob/cairo-2/src/access/accesscontrol/accesscontrol.cairo[{github-icon},role=heading-link]

:assert_only_role: xref:#AccessControl-assert_only_role
:grant_role: xref:#AccessControl-grant_role[grant_role]
:revoke_role: xref:#AccessControl-revoke_role[revoke_role]

```javascript
use openzeppelin::access::accesscontrol::AccessControl;
```

Contract module that allows children to implement role-based access control mechanisms.
Roles are referred to by their `felt252` identifier:

> ```javascript
> // sn_keccak('MY_ROLE')
> const MY_ROLE: felt252 =
>     0x3c877e40edb41710f0baf588c878ee15a04499b06ae8c98cf488875d91a7213;
> ```

Reviewer 3: ```suggestion
```javascript
// 0x3c877e40edb41710f0baf588c878ee15a04499b06ae8c98cf488875d91a7213
const MY_ROLE: felt252 = sn_keccak('MY_ROLE');
```
shouldn't it be like this?

Author: `sn_keccak` is not actually a function in Cairo, but I updated it to use the `selector!` macro.

Reviewer 3: then what we gain by putting `sn_keccak` in there? to me it's confusing rather than clarifying

Author: To me, it shows that the selector macro is returning the sn_keccak. If you think that doesn't add value let me know and will remove it.

Reviewer 3: i do think it's a bit confusing, especially if there's no `sn_keccak` function available, or is it available somewhere else? if we just want to say it's keccak we can simply say that. but I also don't think this is the place to learn how selectors are computed, we just want to compute it. no strong opinion on leaving a comment.

Author: Removed!

* xref:Ownable-OwnershipTransferred[`++OwnershipTransferred(previous_owner, new_owner)++`]
--

[#Ownable-External-Functions]

> ==== External Functions

Reviewer 3: what about putting `transfer_ownership` above? either way we should normalize across the docs to follow the code's definition order as much as possible

Emits an xref:Ownable-OwnershipTransferred[OwnershipTransferred] event.

[.contract-item]
[[Ownable-initializer]]

> ==== `[.contract-item-name]#++initializer++#++(ref self: ContractState, owner: ContractAddress)++` [.item-kind]#internal#

Reviewer 3: we should add `External` and `Internal` titles here to for consistency with ERC20

:RoleRevoked: xref:#IAccessControl-RoleRevoked[RoleRevoked]
:RoleAdminChanged: xref:#IAccessControl-RoleAdminChanged[RoleAdminChanged]

```javascript
use openzeppelin::access::accesscontrol::interface::IAccessControl;
```

External interface of AccessControl.

[.contract-index]
.Interface ID
--
0x23700be02858dbe2ac4dc9c9f66d0b6b0ed81ec7f970ca6844500a56ff61751

> --

Reviewer 3: should we add a link to SRC5's section for anyone who appears in here without enough context? it should be easy to maintain if we're already standardizing this "interface id" section across APIs.

Author: I would add them on the introspection PR, because the section doesn't exist on this one. I added a link just to confirm if you like the approach, let me know (the link is broken here, but should work after merging with the other PR).

:RoleRevoked: xref:#IAccessControl-RoleRevoked[RoleRevoked]
:RoleAdminChanged: xref:#IAccessControl-RoleAdminChanged[RoleAdminChanged]

```javascript
use openzeppelin::access::accesscontrol::interface::IAccessControl;
```

External interface of AccessControl.

[.contract-index]

> .{inner-src5} ID

Reviewer 3: nice! but the formatting looks a bit weird tho
![image](https://github.com/OpenZeppelin/cairo-contracts/assets/1703831/9e66e4a1-516b-47b9-abc2-88c319e2ed5c)
what about incorporating "ID" into the link?

External interface of AccessControl.

[.contract-index]

> .{inner-src5}

Reviewer 3: i think the link is broken, it results in

`https://deploy-preview-719--oz-contracts-cairo.netlify.app/contracts-cairo/0.7.0/api/access#api/introspection.adoc#ISRC5`

Author: This PR doesn't have SRC5, the link will work after merging it with #721

Reviewer 3: right but the link is broken nevertheless. it shouldn't be `api/access#api/introspection.adoc`, or are you saying it is malformed only because it could not resolve the xref?

Author: It works on the introspection PR, right
